// react-sigma/lib/Dagre.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

require("../sigma/layout.dagre");

var _ReactSigmaLayoutPlugin = _interopRequireDefault(require("./ReactSigmaLayoutPlugin"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/**

Dagre layout algorythm.
It supposes that sigma graph is already in place, therefore component should not be
mounted while graph is unavailable. It can be used within Sigma component if graph is
preloaded, or within loader component, like NeoCypher.

It accepts all the parameters of Dagre described on its github page:
@param {boolean}	directed?
@param {boolean}	multigraph?
@param {boolean}	compound?
@param {"TB"|"BT"|"RL"|"LR"}	rankDir?
@param {Sigma$Easing} easing  Easing mode

[see sigma plugin page for more details](https://github.com/Linkurious/linkurious.js/tree/develop/plugins/sigma.layouts.dagre)

**/
const Dagre = props => !!props.sigma ? /*#__PURE__*/_react.default.createElement(_ReactSigmaLayoutPlugin.default, _extends({
  start: () => sigma.layouts.dagre.start(props.sigma),
  config: options => sigma.layouts.dagre.configure(props.sigma, options),
  stop: () => console.warn("dagre stop not implemented")
}, props)) : null;

Dagre.propTypes = {
  directed: _propTypes.default.bool,
  multigraph: _propTypes.default.bool,
  compound: _propTypes.default.bool,
  rankDir: _propTypes.default.oneOf(["TB", "BT", "RL", "LR"]),
  easing: function () {
    return (typeof Sigma$Easing === "function" ? _propTypes.default.instanceOf(Sigma$Easing) : _propTypes.default.any).apply(this, arguments);
  },
  sigma: function () {
    return (typeof Sigma === "function" ? _propTypes.default.instanceOf(Sigma) : _propTypes.default.any).apply(this, arguments);
  }
};
var _default = Dagre;
exports.default = _default;
